### PR TITLE
Do not rematerialize recursively.

### LINKED ISSUE #6731: [Yul] optimizer takes too long for recursive function calls failing with variable x is N slots too deep inside the stack
## Description

```
{
function foo_noreturn_0(x_0, x_1, x_2, x_3)
{
foo_noreturn_0(extcodehash(1), 1, addmod(x_2, x_0, x_3), addmod(x_3, x_3, x_3))
}
foo_noreturn_0(calldataload(0),calldataload(32),calldataload(64),calldataload(96))
}
```

takes a little under two minutes to throw this assertion

https://github.com/ethereum/solidity/blob/e5d46767f1746157be0d346c24a2815d7fdc7306/libyul/backends/evm/EVMCodeTransform.cpp#L806

## Environment

- Compiler version: latest develop

### PR DESCRIPTION
**Pull request by chriseth**
Partly fixes https://github.com/ethereum/solidity/issues/6731

### CODE REVIEW

File: libyul/optimiser/StackCompressor.cpp [hunk 1]
```diff
@@ -123,7 +124,19 @@ void eliminateVariables(Dialect const& _dialect, ASTNode& _node, size_t _numVari
...
+		// If a variable we would like to eliminate references another one
+		// we already selected for elimination, then stop selecting
+		// candidates. If we would add that variable, then the cost calculation
+		// for the previous variable would be off. Furthermore, we
+		// do not skip the variable because it would be better to properly re-compute
+		// the costs of all other variables instead.
+		bool referencesVarToEliminate = false;
+		for (YulString const& referencedVar: get<2>(costs))
+			if (varsToEliminate.count(referencedVar))
+				referencesVarToEliminate = true;
```
Comment (leonardoalt) [inline libyul/optimiser/StackCompressor.cpp:137]: Could also break here

Comment (chriseth) [reply libyul/optimiser/StackCompressor.cpp:137]: Oh right!

**Review by leonardoalt — DISMISSED** [review]

**Comment by bshastry** [comment]
This PR looks like a duplicate of #6820 which has already been merged. 

@chriseth This PR can therefore be closed.

**Comment by chriseth** [comment]
It is not a duplicate but it contains the other one. Will rebase.

**Pull request merged**